### PR TITLE
Make cudnn convolution layer and projection support for dilation

### LINKED ISSUE #3628: Make cudnn convolution layer and projection support for dilation

### PR DESCRIPTION
**Pull request by wanghaoshuang**
fix #3628

### CODE REVIEW

File: paddle/cuda/src/hl_cuda_cudnn.cc [hunk 1]
```diff
@@ -625,18 +633,23 @@ void hl_create_convolution_descriptor(hl_convolution_descriptor* conv,
...
-                                                       1,
-                                                       1,
+                                                       dilation_h,
+                                                       dilation_w,
                                                        mode,
                                                        data_type));
 #else
+  if (dilation_h > 1 || dilation_w > 1) {
+    LOG(FATAL)
+        << "Current cudnn version does't support for dilation convolution.";
```
Comment (qingqing01) [inline paddle/cuda/src/hl_cuda_cudnn.cc]: "Current cuDNN version does't support for dilation convolution.  The dilation convolution requires cuDNN >= v6.0."

File: paddle/gserver/tests/test_LayerGrad.cpp [hunk 2]
```diff
@@ -253,8 +262,8 @@ TEST(Projection, conv) {
   testProjectionConv(1, false);
   testProjectionConv(3, false);
   /// test ConvTransProjection
-  testProjectionConv(1, true);
-  testProjectionConv(3, true);
+  /// testProjectionConv(1, true);
+  /// testProjectionConv(3, true);
```
Comment (qingqing01) [inline paddle/gserver/tests/test_LayerGrad.cpp]: Why comment?

File: python/paddle/trainer/config_parser.py [hunk 3]
```diff
@@ -870,12 +871,15 @@ def __init__(self,
                  caffe_mode=True,
                  filter_size_y=None,
                  padding_y=None,
+                 dilation_y=None,
                  stride_y=None):
```
Comment (qingqing01) [inline python/paddle/trainer/config_parser.py]: It's better to move `dilation=None` and `dilation_y=None` to last line.

File: python/paddle/trainer_config_helpers/layers.py [hunk 4]
```diff
@@ -2458,6 +2465,16 @@ def img_conv_layer(input,
...
 
+    if dilation_y is None:
+        if isinstance(dilation, collections.Sequence):
+            assert len(dilation) == 2
+            dilation, dilation_y = dilation
+        else:
+            dilation_y = dilation
+
+    if dilation > 1 or dilation_y > 1:
+        assert layer_type in ["cudnn_conv", "cudnn_convt"]
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: If the `layer_type =None`, this line will fail.

File: python/paddle/trainer_config_helpers/layers.py [hunk 5]
```diff
@@ -2340,13 +2340,15 @@ def img_conv_layer(input,
                    groups=1,
                    stride=1,
                    padding=0,
+                   dilation=0,
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: 1 by defalut.

File: paddle/gserver/tests/test_LayerGrad.cpp (hunk 2 above)
Comment (wanghaoshuang) [inline paddle/gserver/tests/test_LayerGrad.cpp]: Fixed.

File: python/paddle/trainer/config_parser.py (hunk 3 above)
Comment (wanghaoshuang) [inline python/paddle/trainer/config_parser.py]: Fixed.

File: python/paddle/trainer_config_helpers/layers.py (hunk 5 above)
Comment (wanghaoshuang) [inline python/paddle/trainer_config_helpers/layers.py]: Fixed.

File: python/paddle/trainer_config_helpers/layers.py (hunk 4 above)
Comment (wanghaoshuang) [inline python/paddle/trainer_config_helpers/layers.py]: Fixed.

**Pull request merged**